include/lo2s/types/fd.hpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2017,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+extern "C"
+{
+#include <unistd.h>
+}
+
+namespace lo2s
+{
+
+class WeakFd;
+class Fd
+{
+public:
+    explicit Fd(int fd) : fd_(fd)
+    {
+    }
+
+    friend WeakFd;
+
+    Fd(Fd&) = delete;
+    Fd& operator=(Fd&) = delete;
+
+    Fd(Fd&& other)
+    {
+        fd_ = other.fd_;
+        other.fd_ = -1;
+    }
+
+    Fd& operator=(Fd&& other)
+    {
+        fd_ = other.fd_;
+        other.fd_ = -1;
+        return *this;
+    }
+
+    int as_int() const
```

Comment:
The `as_int` seems pretty obnoxious in code, maybe define the `int operator*()`?